### PR TITLE
DHQ Recommendation Box for Static Site Generation

### PR DESCRIPTION
**Pull request by joelsjlee**
This PR is to add the recommendations box at the end of article to the DHQ website. It takes in three tsv files which I have added in a subfolder `data/dhq-recs`, and runs xslt code (`dhq2html.xsl`) over the files to generate the top 5 recommendations for each of the recommendation systems. I have edited minimally the css to include these recommendations in the same styling as the Abstract and I have implemented the namespace in the `dhq2html.xsl` file to use for a function.

This PR also merges into it Julia's `add_recommendations` branch.

A couple things left to still do:

- [x] Connect this PR with Julia's `add_recommendations` branch, and edit the recommendation text inside the box to link out to the `explore.html` page.
- [x] Take a look at the styling used for box and whether or not it could be improved for accessibility (especially the coloring of the links).
- [x] Check in with Haining and Ben about final TSV versions and any discrepencies
- [x] Any further testing

Also, this pull request is being done from `static_site_generation` into `static_site_for-merging` because of a mishap from my part, shoutout and big thanks to Ash for coming up with a solution for this.

### CODE REVIEW

File: common/xslt/dhq2html.xsl [hunk 1]
```diff
@@ -237,6 +239,171 @@
...
+                    <h3>SPECTER Recommendations</h3>
+                    <p>Below are article recommendations generated by the SPECTER model:</p>
+                    <ol>
+                        <li>
+                            <!-- create variable to pass as parameter. This is index 6 of the tsv, or the first recommended article ID -->
+                            <xsl:variable name="input_num" select="6" />
+                            <!-- Calling the function to populate the recommendation. Pass in the input number, the current row of the selected Article,
+                            and the whole set of rows in the tsv. -->
+                            <xsl:sequence select="dhqf:recs($input_num, $spector_tabs, $spector_rows)" />
+                        </li>
```
Comment (amclark42) [inline common/xslt/dhq2html.xsl:308]: Here, we could actually use an `<xsl:for-each>` to remove some of the duplicated code in this list. `<xsl:for-each>` doesn't _just_ work on XML nodes or a pre-generated sequence of things. You can also hand it a range of numbers (TSV indexes? 👀) and iterate over each one.

The code below will output 6 messages to the console: 1, 2, 3, 4, 5, 6
```xslt
<xsl:for-each select="1 to 6">
  <xsl:message select="."/>
</xsl:for-each>
```

So, we can replace all those `<li>`s with a single `<xsl:for-each>`:
```xslt
                        <!-- The recommendations for this article start at column 6 of the TSV and end at column 10 -->
                        <xsl:for-each select="6 to 10">
                          <li>
                            <!-- create variable to pass as parameter. This is the index of a recommended article ID -->
                            <xsl:variable name="input_num" select="." />
                            <!-- Calling the function to populate the recommendation. Pass in the input number, the current row of the selected Article,
                            and the whole set of rows in the tsv. -->
                            <xsl:sequence select="dhqf:recs($input_num, $spector_tabs, $spector_rows)" />
                          </li>
                        </xsl:for-each>
```

Given that this code is also repeated for each of the three sections, we could further abstract this code out by making it a function. It'd only need the current row's cells and the other rows:

```xslt
    <xsl:function name="dhqf:make-recs-list-items">
      <xsl:param name="article-row-cells"/>
      <xsl:param name="all-rows"/>
      <!-- The recommendations for this article start at column 6 of the TSV and end at column 10. -->
      <ol>
        <xsl:for-each select="6 to 10">
          <li>
            <!-- create variable to pass as parameter. This is the index of a recommended article ID -->
            <xsl:variable name="input_num" select="." />
            <!-- Calling the function to populate the recommendation. Pass in the input number, the current row of the selected Article,
              and the whole set of rows in the tsv. -->
            <xsl:sequence select="dhqf:recs($input_num, $article-row-cells, $all-rows)" />
          </li>
        </xsl:for-each>
      </ol>
    </xsl:function>
```
We could then call that function for each set of recommendations, e.g.
```xslt
                    <p>Below are article recommendations generated by the SPECTER model:</p>
                    <xsl:sequence select="dhqf:make-recs-list($spector_tabs, $spector_rows)"/>
```

@joelsjlee, I've got a partially-working solution for this too. Would you like me to commit it, save you some time?

**Review by amclark42 — APPROVED** [review]
@joelsjlee Overall, this is very solid work! I have a few suggestions and recommendations, which are given in the comments above. I have code for two of the recommendations in `dhq2html.xsl`, which I'm happy to commit and save you some work. I can also implement the CSS changes I suggested if you'd like.

File: common/xslt/dhq2html.xsl (hunk 1 above)
Comment (amclark42) [reply common/xslt/dhq2html.xsl:308]: There's an inconsistency in the code I provided: I changed the function name to `dhqf:make-recs-list()` and used that in the `<xsl:sequence>`, but didn't update the function name in my comment! My bad.

Comment (joelsjlee) [reply common/xslt/dhq2html.xsl:308]: Wow again! I should've figured that there was some way to do this in XSLT. Thanks for all the help and yes I would appreciate if you can go ahead and commit the XSLT and CSS changes.

File: common/xslt/dhq2html.xsl [hunk 2]
```diff
@@ -237,6 +239,171 @@
...
                 title="Click for print friendly version">Print</a>
+            |&#x00a0;
+            <a href="#recommendations" onclick="document.getElementById('recommendations').scrollIntoView();"
+                title="See Recommendations">See Recommendations</a>
+        </div>
+    </xsl:template>
+    <!-- Load tsvs -->
+    <xsl:variable name="specter-recs-tsv">
+      <xsl:sequence select="unparsed-text('../../data/dhq-recs/dhq-recs-zfill-spctr.tsv')"/>
+    </xsl:variable>
```
[the remark this answers is not in the document]
Comment (amclark42) [reply common/xslt/dhq2html.xsl:250]: Right, when it has `@select` on it, `<xsl:variable>` acts like `<xsl:sequence>`. It's a useful shorthand!

File: common/xslt/dhq2html.xsl (hunk 1 above)
Comment (amclark42) [reply common/xslt/dhq2html.xsl:308]: On it!

**Comment by amclark42** [comment]
@joelsjlee I've pushed up my changes! Take a look and make any tweaks. You can make the merge whenever you're ready.

**Comment by joelsjlee** [comment]
Everything here looks great! One last thing I forgot to mention before I merge, when I'm generating the site and testing it out on a local server, clicking on the explore.html page that Julia set up gives a 404 (and subsequently the link to it in the recommendation box does the same). I'm assuming this is because the paths aren't set up yet in the ant build file (?). Should we add this to this pull request?

**Comment by juliaflanders** [comment]
That would be my guess--if so, yes, it sounds like adding this to the PR would make sense!

best, Julia

**Comment by amclark42** [comment]
Ah, I didn't notice the page was in a new directory. Let me fix the Ant build...

**Comment by amclark42** [comment]
Done!

**Comment by amclark42** [comment]
@joelsjlee This looks great! Are there articles I can test this on, to see the conditionality in action?

**Comment by joelsjlee** [comment]
Yes! I believe 000672 should have SPECTER recommendations omitted, and 000580 should have all of the recommendations omitted.

**Comment by joelsjlee** [comment]
I'm waiting on the last up-to-date version of the TSVs that removes some of the purposefully omitted articles from being recommended. Once that is all squared away and the TSVs are passing all of our tests, (hopefully by this week) everything should be all set.

**Comment by amclark42** [comment]
@joelsjlee Just tested the new code and it looks great! There's only one small problem: When the Recommendations box is omitted, the toolbars at the top and bottom of the page still show the "See Recommendations" link. Could you fix that?

**Comment by joelsjlee** [comment]
I've added a rudimentary fix... I couldn't figure out if it was possible to have some kind of interaction between the toolbar and recommendation templates to get that part removed within the existing conditional in the recommendation box so I ended up copying and pasting some code. Is there a more elegant solution here?

**Comment by amclark42** [comment]
Good question! I just pushed up an attempt, but don't have it working quite yet. I added a global variable to say, true or false, if the current article has any recommendation in any TSV. It works a bit too well at excluding the Recommendations box and links, though — they're not showing up anymore. (I probably messed up the test.)

**Comment by amclark42** [comment]
I think I got it! Needed to put the "m" flag on `matches()` so each row counts as a new line-beginning (`^`).

**Comment by joelsjlee** [comment]
This looks great! Thanks so much for the fix Ash. I just updated to a stable version of the TSVs. There was a bit of debugging to do, and removing some articles, but everything should be good now. Shall I go ahead and merge the PR?

**Comment by amclark42** [comment]
Yeah, go for it! 🎉

**Pull request merged**